### PR TITLE
Update Japan holidays

### PR DESCRIPTION
**Pull request by KJhellico**
## Proposed change

Update Japan holidays:
- refactor substitute and national holidays calculation
- add Bank Holiday on Jan 1
- update tests

## Type of change

- [ ] New country/market holidays support (thank you!)
- [x] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/pin/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've run `make pre-commit`, it didn't generate any changes
- [x] I've run `make test`, all tests passed locally

[contributing-guidelines]: https://github.com/vacanza/python-holidays/blob/dev/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/python-holidays/tree/dev/docs/source

### CODE REVIEW

File: holidays/countries/japan.py [hunk 1]
```diff
@@ -51,29 +51,34 @@ def __init__(self, *args, **kwargs) -> None:
...
+            self._add_observed(dt, name=tr("振替休日"), show_observed_label=False)
+
+        # A weekday between national holidays becomes a holiday too (国民の休日) -
+        # national holiday.
+        # In 1986-2006 it was only May 4 (between Constitution Day and Children's Day).
+        # Since 2006, it may be only the day between Respect for the Aged Day and
+        # Autumnal Equinox Day (in September).
+        if self._year <= 1985:
+            return None
+        if self._year <= 2006:
```
Comment (PPsyrius) [inline holidays/countries/japan.py:71]:
```suggestion
        elif 1986 <= self._year <= 2006:
```

Comment (KJhellico) [reply holidays/countries/japan.py:71]: But what will we gain from it? 👀

Comment (PPsyrius) [reply holidays/countries/japan.py:71]: Nvm, I thought it was supposed to be part of the same if-else as this section above:
```python
        if self._year <= 1985:
            return None
        if self._year <= 2006:
            may_4 = (MAY, 4)
            if not (self._is_monday(may_4) or self._is_sunday(may_4)):
                # National Holiday.
                self._add_holiday(tr("国民の休日"), may_4)
        else:
            for dt in dts:
                if dt.month == SEP and _timedelta(dt, +2) in dts:
                    # National Holiday.
                    self._add_holiday(tr("国民の休日"), _timedelta(dt, +1))
                    break
```

File: holidays/countries/japan.py [hunk 2]
```diff
@@ -51,29 +51,34 @@ def __init__(self, *args, **kwargs) -> None:
...
+            may_4 = (MAY, 4)
+            if not (self._is_monday(may_4) or self._is_sunday(may_4)):
```
Comment (arkid15r) [inline holidays/countries/japan.py:73]: up to you: this version might be a bit more performant and (personally for me) easier to read.
```suggestion
            if not self._is_monday(may_4) and not self._is_sunday(may_4):
```

**Review by arkid15r — DISMISSED** [review]
LGTM

**Review by arkid15r — APPROVED** [review]

**Pull request merged**